File: HW2/frameZeroJustify.m
function frameMat2 = frameZeroJustify(frameMat, polyOrder)
% frameZeroJustify Zero justification via polynomial fitting
%   frameMat2 = frameZeroJustify(frameMat, polyOrder)
%
%   說明：
%   - frameMat: 輸入的音框矩陣，每一「欄」(column)代表一個音框
%   - polyOrder: 多項式的階數 (例如 3)
%   - frameMat2: 輸出的音框矩陣，已做「零漂移修正」
%
%   每個音框會個別進行：
%   1. 建立時間軸 (或索引) 並做 z-normalization
%   2. 使用 polyfit / polyval 擬合出基線
%   3. 將基線從原音框訊號中扣除，得到去除漂移後的訊號

    [numSamples, numFrames] = size(frameMat);  % 取得每框取樣點數及框數
    frameMat2 = zeros(numSamples, numFrames);  % 初始化輸出矩陣

    % 建立並 z-normalize 時間軸 (索引)
    n = (0:numSamples-1)';               % 時間索引：0, 1, 2, ..., numSamples-1
    n_norm = (n - mean(n)) / std(n);     % z-normalization，避免數值不穩定

    for i = 1:numFrames
        % 取出第 i 欄 (第 i 個音框)
        frameData = frameMat(:, i);

        % 多項式擬合
        p = polyfit(n_norm, frameData, polyOrder);
        baseline = polyval(p, n_norm);

        % 扣除基線得到零漂移修正後的音框
        frameMat2(:, i) = frameData - baseline;
    end
end

File: HW2/frameZeroJustify.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile

if __name__ == "__main__":
    sample_rate, raw_data = wavfile.read("star_noisy.wav")
    raw_data = raw_data[0:16000]
    
    fig, ax = plt.subplots(3, 1)
    fig.set_figwidth(10)
    fig.set_figheight(8)
    fig.canvas.manager.set_window_title("D1009212 - HW02")
    plt.subplots_adjust(hspace=0.5)
    
    # (a) Waveform of the star_noisy.wav
    norm_data = raw_data / np.max(np.abs(raw_data))
    norm_data = norm_data - np.mean(norm_data)
    plt.subplot(3, 1, 1)
    plt.plot(norm_data, linewidth=0.5, color='blue')
    plt.axvline(x=6000, color='red', linewidth=0.5)
    plt.axvline(x=6256, color='red', linewidth=0.5)
    plt.grid(linestyle='--')
    plt.title("(a) Waveform of the star_noisy.wav")
    plt.xlim(0, 16000)
    plt.ylim(-1, 1)
    
    # (b) Frame (blue) and the fitted line of third-order polynomial (red)
    frame_data = norm_data[6000:6000 + 256]
    x = np.arange(256)
    coefs = np.polyfit(x, frame_data, 3)
    coefs_point = np.polyval(coefs, x)
    plt.subplot(3, 1, 2)
    plt.plot(frame_data, linewidth=1, color='blue')
    plt.plot(coefs_point, linewidth=1, color='red')
    plt.grid(linestyle='--')
    plt.title("(b) Frame (blue) and the fitted line of third-order polynomial (red)")
    plt.xlim(0, 256)
    
    # (c) Frame after zero-justification
    zj_frame_data = frame_data - coefs_point
    plt.subplot(3, 1, 3)
    plt.plot(zj_frame_data, linewidth=1, color='blue')
    plt.grid(linestyle='--')
    plt.title("(c) Frame after zero-justification")
    plt.xlim(0, 256)
    
    plt.show()